cli/routing.go

```diff
@@ -0,0 +1,108 @@
+// Copyright 2018 Prometheus Team
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cli
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+
+	"github.com/xlab/treeprint"
+
+	"github.com/prometheus/alertmanager/client"
+	"github.com/prometheus/alertmanager/dispatch"
+	"gopkg.in/alecthomas/kingpin.v2"
+)
+
+type routingShow struct {
+	configFile        string
+	labels            []string
+	expectedReceivers string
+	tree              treeprint.Tree
+	debugTree         bool
+}
+
+const routingHelp = `Prints alert routing tree
+
+Will print whole routing tree in form of ASCII tree view.
+
+Routing is loaded from configuration file or remote Alertmanager.
+Use --config.file or --alertmanager.url flags to specify where to look for the configuration.
+
+Example:
+
+./amtool config routes [show] --config.file=doc/examples/simple.yml
+
+`
+
+func configureRoutingCmd(app *kingpin.CmdClause) {
+	var (
+		c              = &routingShow{}
+		routingCmd     = app.Command("routes", routingHelp)
+		routingShowCmd = routingCmd.Command("show", routingHelp).Default()
+	)
+	routingCmd.Flag("config.file", "Config file to be tested").ExistingFileVar(&c.configFile)
+	routingShowCmd.Action(execWithTimeout(c.routingShowAction))
+	configureRoutingTestCmd(routingCmd, c)
+}
+
+func (c *routingShow) routingShowAction(ctx context.Context, _ *kingpin.ParseContext) error {
+
+	// Load configuration form file or URL.
+	cfg, err := loadAlertmanagerConfig(ctx, alertmanagerURL, c.configFile)
+	if err != nil {
+		kingpin.Fatalf("%s", err)
+		return err
+	}
+	route := dispatch.NewRoute(cfg.Route, nil)
+	tree := treeprint.New()
+	convertRouteToTree(route, tree)
+	fmt.Println("Routing tree:")
+	fmt.Println(tree.String())
+	return nil
+}
+
+func checkRoutingConfigInputFlags(alertmanagerURL *url.URL, configFile string) {
+	if alertmanagerURL != nil && configFile != "" {
+		kingpin.Fatalf("You can't use both --config.file and --alertmanager.url at the same time.")
+	}
+	if alertmanagerURL == nil && configFile == "" {
+		kingpin.Fatalf("You have to specify one of --config.file or --alertmanager.url flags.")
+	}
+}
+
+func convertRouteToTree(route *dispatch.Route, tree treeprint.Tree) {
+	routeDecription := fmt.Sprintf("%s  continue: %t  receiver: %s", route.Matchers.String(), route.Continue, route.RouteOpts.Receiver)
```

Comment:
When I run this against SoundCloud's config (which has many many routes), it ends up being fairly noisy. I'm willing to bet most people don't ever use `continue`, so how about we only mention continue if its value is set to true?

Also, I'm thinking about maybe only showing the receiver when a user uses `show`. When using `test`, the matches get written out at the bottom so maybe it's not necessary?

Reply:
>About the test and showing a receiver in the tree. I think it doesn't hurt and it's easier to see from which branch comes which receiver (even though it matches all leafs top-down to receivers left-to-right). But if you want I can remove it.

We can leave it in for now and think about it a bit more.

>Could you possibly share how the output looks like for such complex configuration? Just out of curiosity I was lazy to write more complex routing tree :)

It's a lot to sanitize, I'll figure out a way to get it to you. I'm on vacation for the next few days so it'll be early next week.

Reply:
Forget it if you'd need to sanitize it, I was just curious :)